File: zig/src/tools/eval_provider_factory.zig
//! eval-provider-factory — trimmed provider factory parity runner.

const std = @import("std");
const zeroclaw = @import("zeroclaw");
const factory = zeroclaw.providers.factory;

const EvalError = error{InvalidScenario};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout);
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();

    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;
    if (!std.mem.eql(u8, op, "create")) return EvalError.InvalidScenario;

    const name = getString(parsed.value, "name") orelse return EvalError.InvalidScenario;
    const api_key = getOptionalString(parsed.value, "api_key");
    const url = getOptionalString(parsed.value, "url");

    if (factory.createProviderWithUrl(allocator, name, api_key, url)) |handle_value| {
        var handle = handle_value;
        defer handle.deinit(allocator);
        try writer.writeAll("{\"op\":\"create\",\"result\":{\"ok\":true,\"provider_name\":");
        try std.json.stringify(handle.providerName(), .{}, writer);
        try writer.writeAll("}}\n");
    } else |err| {
        try writer.writeAll("{\"op\":\"create\",\"result\":{\"ok\":false,\"error\":");
        try std.json.stringify(errorTag(err), .{}, writer);
        try writer.writeAll("}}\n");
    }
}

fn errorTag(err: anyerror) []const u8 {
    return switch (err) {
        factory.FactoryError.ProviderNotSupported => "provider_not_supported",
        factory.FactoryError.ApiKeyPrefixMismatch => "api_key_prefix_mismatch",
        factory.FactoryError.MissingApiKey => "missing_api_key",
        else => "out_of_memory",
    };
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .string) return null;
    return inner.string;
}

fn getOptionalString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    return switch (inner) {
        .string => inner.string,
        .null => null,
        else => null,
    };
}

File: eval-tools/src/bin/eval-provider-factory.rs
//! eval-provider-factory — trimmed provider factory parity runner.

use std::io::{self, Read, Write};

use anyhow::{Context, Result};
use serde_json::Value;
use zeroclaw_providers::create_provider_with_url;

fn main() -> Result<()> {
    // Keep the eval deterministic even when the developer shell has provider
    // credentials configured. The parent process is unaffected.
    for key in [
        "OPENAI_API_KEY",
        "ZEROCLAW_API_KEY",
        "API_KEY",
        "ZEROCLAW_PROVIDER_URL",
    ] {
        std::env::remove_var(key);
    }

    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for raw_line in input.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = required_string(&value, "op")?;
        run_op(&mut output, op, &value)?;
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

fn run_op(output: &mut Vec<u8>, op: &str, value: &Value) -> Result<()> {
    match op {
        "create" => {
            let name = required_string(value, "name")?;
            let api_key = optional_string(value, "api_key");
            let url = optional_string(value, "url");
            write_create_result(output, name, create_trimmed(name, api_key, url));
        }
        _ => anyhow::bail!("unknown provider_factory op: {op}"),
    }
    Ok(())
}

fn create_trimmed(
    name: &str,
    api_key: Option<&str>,
    url: Option<&str>,
) -> Result<(), &'static str> {
    match name {
        "openai" | "ollama" => create_provider_with_url(name, api_key, url)
            .map(|_| ())
            .map_err(|err| canonical_error_tag(&err.to_string())),
        _ => Err("provider_not_supported"),
    }
}

fn canonical_error_tag(message: &str) -> &'static str {
    if message.contains("API key prefix mismatch") {
        "api_key_prefix_mismatch"
    } else {
        "provider_not_supported"
    }
}

fn write_create_result(
    output: &mut Vec<u8>,
    provider_name: &str,
    result: Result<(), &'static str>,
) {
    let value = match result {
        Ok(()) => serde_json::json!({
            "op": "create",
            "result": { "ok": true, "provider_name": provider_name },
        }),
        Err(error) => serde_json::json!({
            "op": "create",
            "result": { "ok": false, "error": error },
        }),
    };
    serde_json::to_writer(&mut *output, &value).expect("serialize eval result");
    output.push(b'\n');
}

fn required_string<'a>(value: &'a Value, key: &str) -> Result<&'a str> {
    value
        .get(key)
        .and_then(Value::as_str)
        .with_context(|| format!("{key} missing"))
}

fn optional_string<'a>(value: &'a Value, key: &str) -> Option<&'a str> {
    value.get(key).and_then(Value::as_str)
}
